File: notes/2021/10/07/point.py
import math


class Point:
    def __init__(self, x: int = 0, y: int = 0):
        self.x = x
        self.y = y

    def distance_to(self, q) -> float:
        return math.hypot(self.x - q.x, self.y - q.y)
    
    def reflect_across_x(self):
        return Point(self.x, -self.y)
    
    def reflect_across_y(self):
        return Point(-self.x, self.y)
    
    def reflect_across_yx(self):
        return Point(self.y, self.x)

    def __repr__(self):

        return f"<Point x={self.x}, y={self.y}>"
    
    def __eq__(self, o):
        if type(o) != Point:
            return False
        if o is self:
            return True
        return self.x == o.x and self.y == o.y


if __name__ == "__main__":

    point = Point(3, 4)
    point_2 = Point()
    print(point.distance_to(point_2))

File: notes/2021/10/07/Point.java
public class Point 
{
    private final int x; // final means not re-assignable
    private final int y;

    public Point(int x, int y) 
    {
        this.x = x;
        this.y = y;
    }

    public Point() // method name overloading
    {
        this(0, 0); // Calls sibling constructor above
    }

    public double distanceTo(Point q) {
        return Math.hypot(x - q.x, y - q.y);
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Point reflectionAcrossX()
    {
        return new Point(x, -y);
    }

    public Point reflectionAcrossY()
    {
        return new Point(-x, y);
    }

    public Point reflectionAcrossYX()
    {
        return new Point(y, x);
    }

    @Override
    public String toString() {
        return String.format("<Point x=%s, y=%s>", x, y);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Point)) {
            return false;
        }
        if (this == obj) {
            return true;
        }
        Point p = (Point) obj;
        return p.x == x && p.y == y;

    }

    public static void main(String[] args) {
        Point p = new Point(3, 4);
        Point q = new Point();
        System.out.println(p.distanceTo(q));
    }
}
